Add design space guidelines

barklund opened the pull request:
## Summary

This fixes the general guidelines in a few ways and then adds the design space guideline.

Demo:
![design-space](https://user-images.githubusercontent.com/637548/94324793-7eaae780-ff69-11ea-8ff6-bd8c656655f8.gif)

## Relevant Technical Choices

* Many

## To-do

- [x] Change color of guidelines
- [x] Don't snap to the background element
- [x] Move general guidelines to correct offset
- [x] Add rounding for general guidelines (kind of)
- [x] Add design space guideline element and show if snapping to it
- [x] Fix bug around design space guidelines not being hidden when snapped to and the element is dropped
- [x] Needs visual karma test

## User-facing changes

There's a new design space guideline to snap to - also snaplines are improved in general.

## Testing Instructions

<!-- How can the changes in this PR be verified? Relevant for QA  and user acceptance testing. -->

---

<!-- Please reference the issue(s) this PR addresses. -->

Fixes #2917

Commit:
Updated snap guideline color

Commit:
Attempting to fix guidelines

Commit:
Merge branch 'main' into add/2917-recommended-design-space

Commit:
Excluded bg elements from other node guidelines in moveable

Commit:
Fixed offset for snapping guidelines

Commit:
Snap to zero when close on drag - not perfect but better

Commit:
Added design space guidelines

miina commented on assets/src/edit-story/components/canvas/utils/useSnapping.js:
```diff
@@ -14,34 +14,100 @@
+  const offsetY = Math.floor(pageRect.y - canvasRect.y);
+
+  const verticalGuidelines = canSnap
+    ? [offsetX, offsetX + canvasWidth / 2, offsetX + canvasWidth]
```
This seems to be based on the safezone but not the fullbleed -- meaning that the snaplines would shown only on the safezone, shouldn't the fullbleed snaplines still remain as well for being able to snap to the edges? Especially when the `showSafeZone` is set to `false`? Couldn't we use `fullbleedContainer` here?

miina replied:
Hmm, generally curious why the vertical / horizontal guidelines needed changing.

barklund replied:
Because they snapped to the top left corner of the workspace! They've been broken for a long time - it just wasn't noticed because we also had the bg element as a snap target.

What else would these weird phantom snaplines be (deus ex machina)?

<img width="595" alt="Screen Shot 2020-09-28 at 11 48 23" src="https://user-images.githubusercontent.com/637548/94455121-b8097000-0180-11eb-940b-b3df7c718e6d.png">

And yes, it makes sense to snap to the fullbleed element if safe mode is off.

miina replied:
> Because they snapped to the top left corner of the workspace! 

Hah, okay! I thought the PR changed the snapping from full-bleed to safe zone only but if it was snapping to the workspace initially then that's another story.

miina commented on assets/src/edit-story/components/canvas/utils/useSnapping.js:
```diff
@@ -14,34 +14,100 @@
+  const handleSnap = useCallback(
+    ({ elements }) =>
+      // Show design space if we're snapping to any of its edges
+      toggleDesignSpace(
```
When testing, I saw the design space snaplines being shown while drop targeting as well. Assuming this is part of one of the todo-s? Or perhaps that was happening already before?
![droptargeting](https://user-images.githubusercontent.com/3294597/94432479-1e00f200-0197-11eb-84db-36d86c824c39.gif)

barklund replied:
Hm, all snapping guidelines should really be disabled - you also see the guideline for page center there, which is also misleading. Can we do that dynamically during drag?

miina replied:
Perhaps if there's an `activeDropTargetId` set, snapping could be disabled?

barklund replied:
Hey, that worked surprisingly well, thanks! Done in 17e9d47a6.

miina commented on assets/src/edit-story/components/canvas/utils/useSnapping.js:
```diff
@@ -14,34 +14,100 @@
+    : [];
+
+  const elementGuidelines = canSnap
+    ? [...otherNodes, ...(showSafeZone ? [designSpaceGuideline] : [])]
```
Good idea adding it to the elements list.

miina replied:
One thing I'm wondering about is if the design space should snap even if the element is touching it from the outside only (while being fully outside of the design space), this kind of defeats the purpose of keeping the elements inside the space. However, perhaps it's OK for now as-is. WDYT?

barklund replied:
But how would we do that? Yes, we can not show the design space guideline (the pink box) if it touches from the outside, but moveable would still show the vertical snapping guideline - we can't add a guideline or snapping element and say "only snap if on the left side" or something like that. So, we'd be showing vertical snapping guidelines in weird places, where it seems like there's nothing to snap to.

barklund replied:
Maybe our good friend @daybrush can help us out here.

miina replied:
Hmm, I guess we could do the calculation during transformation to understand if the element is outside of the design space, and if it is, disable the relevant snapping guidelines by passing in only `otherNodes`.

Anyway, this behavior is not specified by UX currently so I think it's OK to leave it for the future if it becomes relevant.

barklund replied:
Hm, yeah, that's true - we could disable that as a snapping element as soon as the selection left boundary is over the left border or the selection right boundary is over the right border. But that's pretty complex. I'll leave that for now.

miina commented on assets/src/edit-story/components/canvas/singleSelectionMoveable/useDrag.js:
```diff
@@ -87,12 +87,15 @@ function useSingleSelectionDrag({
       return undefined;
     }
+
+    const roundToZero = (num) => (Math.abs(num) <= 1 ? 0 : num);
```
This name is perhaps a bit confusing -- not sure of a better one either though.

When does this happen -- does snapping to the design space cause numbers that are not round?

barklund replied:
Sometimes it's 1, sometimes it's -1. It's because we do a lot of rounding and it's often off. It's still off - try snapping to the 48px design guideline - often the x value of the "snapped" element will be 49.

It's an underlying bug. We should fix it by doing something like this:

* If you drop an object while it's snapping to something, use the coordinate of what it's snapping to rather than the regular "initial + delta" calculation.

barklund replied:
See this, which also occurs in main:

![snap-rounding-off](https://user-images.githubusercontent.com/637548/94493388-75648980-01ba-11eb-9987-1fc3a973c9f7.gif)

Even though the dropped element seems to snap to the left side of an element at 100px, the dropped element x can easily vary by 1px. And actually does very often - more than not!

miina replied:
Hmm, okay. Maybe let's just create a follow-up issue for now?

barklund replied:
Added in #4813.

miina commented on assets/src/edit-story/components/canvas/framesLayer.js:
```diff
@@ -67,6 +67,18 @@ const Hint = styled.div`
+const DesignSpaceGuideline = styled.div`
+  border: 1px solid ${({ theme }) => theme.colors.callout};
+  left: ${100 * (48 / 412)}%;
+  right: ${100 * (48 / 412)}%;
```
Let's use constants here for better readability.

barklund replied:
Done in 6eaf4e4e1.

Commit:
Never snap if on a droptarget

Commit:
Added constants for margin ratio

Commit:
Hiding design space when not dragging

Also fixing a bug with `isDragging` never being true for single selection moveable - was not really used, but is now simplifying some other stuff.

Commit:
Merge branch 'main' into add/2917-recommended-design-space

Commit:
Made the edits more test-proof

miina reviewed (approved):
LGTM!

Commit:
Added visual karma test for design space guideline

Commit:
Merge branch 'main' into add/2917-recommended-design-space

Commit:
Merge branch 'main' into add/2917-recommended-design-space
